OCPBUGS-29510: Fix console crash-loop with OIDC auth config

TheRealJon opened the pull request:
When the "user-auth" flag is set to disabled, the web console should display a message indicating that authentication is required for the console to function, with the exception where a static user configuration is being used.

TheRealJon commented:
/jira refresh

stlaz reviewed (changes requested):
From what I understand, the "disabled" handler might be used for testing/local dev?

I guess we'll need a new "nothing-configured" state. Ideally we might want to replace `srv.Authenticator` in "disabled" case with something that can later be used in the authentication middleware.

That should allows us to distinguish a case where authenticator was not configured (authType == "none" and the `srv.Authenticator == nil` in that case, maybe?) compared to where it should just be a SA token passthrough.

We could then reuse the `Authenticator == nil` logic in https://github.com/openshift/console/blob/537716b67aa4e294ad7d44034512efabb40a59bd/pkg/server/server.go#L281-L298 ?

TheRealJon commented:
Hmm, yeah, we might need to detect the case where "user-auth" is disabled, but still allow the console to run when a static user is defined.

TheRealJon commented:
@stlaz @jhadvig I've updated. Console should work when running with a static user now.

stlaz commented:
/hold
I want to take some time to think about this a bit more

yanpzhan commented:
@TheRealJon @stlaz  The bug is set as release blocker, if there will be update for the pr?

stlaz commented on cmd/bridge/main.go:
```diff
@@ -528,7 +528,7 @@ func main() {
 		srv.InternalProxiedK8SClientConfig.BearerToken = *fK8sAuthBearerToken
 	case "oidc", "openshift":
-		flags.FatalIfFailed(flags.ValidateFlagIs("user-auth", authOptions.AuthType, "oidc", "openshift"))
+		flags.FatalIfFailed(flags.ValidateFlagIs("user-auth", authOptions.AuthType, "oidc", "openshift", "disabled"))
```
To be honest, I believe the whole `k8s-auth` flag should be removed. The only value that appears to make sense is `bearer-token`, which in fact replaces your authenticator. That should be its own flag that basically makes the console operate in a non-proxy mode with a hardcoded test-user identified by the token.
From that point, you could just create your own authenticator that uses the hardcoded user (you'll have to change the authenticator struct to interface).
When you get here, you'll be able to determine in the `srv.HTTPHandler()` that you should return a handler that sends you a "No authentication configured" web page.

TheRealJon replied:
We don't have time to make this big of a change now. This would also change the API, which is out of scope for a bug fix. We could follow up with a tech debt story though.

TheRealJon replied:
I mean removing the `k8s-auth` flag, not the suggested change.

stlaz replied:
There will never be time, but I can do it if you want.

TheRealJon commented:
@yapei No more updates. I just had to rebase.

yanpzhan commented:
@TheRealJon @stlaz  will the fix for the bug be available for 4.16.0?

stlaz commented:
It should be, I've already started working on it.

TheRealJon commented:
@stlaz @jhadvig I've made some changes, PTAL.

@standa I incorporated the changes from [your WIP PR](https://github.com/openshift/console/pull/13802/files) and added a few tweaks.

TheRealJon commented:
/label tide/merge-method-squash

stlaz commented on cmd/bridge/config/auth/authoptions.go:
```diff
@@ -67,13 +67,13 @@ func NewAuthOptions() *AuthOptions {
 func (c *AuthOptions) AddFlags(fs *flag.FlagSet) {
 	fs.StringVar(&c.AuthType, "user-auth", "", "User authentication provider type. Possible values: disabled, oidc, openshift. Defaults to 'openshift'")
-	fs.StringVar(&c.IssuerURL, "user-auth-oidc-issuer-url", "", "The OIDC/OAuth2 issuer URL.")
+	fs.StringVar(&c.IssuerURL, "user-auth-oidc-issuer-url", "", "The OIDC OAuth2 issuer URL.")
 	fs.StringVar(&c.ClientID, "user-auth-oidc-client-id", "", "The OIDC OAuth2 Client ID.")
 	fs.StringVar(&c.ClientSecret, "user-auth-oidc-client-secret", "", "The OIDC OAuth2 Client Secret.")
 	fs.StringVar(&c.ClientSecretFilePath, "user-auth-oidc-client-secret-file", "", "File containing the OIDC OAuth2 Client Secret.")
-	fs.StringVar(&c.CAFilePath, "user-auth-oidc-ca-file", "", "Path to a PEM file for the OIDC/OAuth2 issuer CA.")
```
OIDC and OAuth2 are different protocols, I think we should keep the slash

stlaz commented on pkg/serverconfig/config.go:
```diff
@@ -233,14 +232,6 @@ func addClusterInfo(fs *flag.FlagSet, clusterInfo *ClusterInfo) {
 	}
 }
 
-func defaultK8SAuth(fs *flag.FlagSet) {
```
:heart: I totally forgot about this, thanks

stlaz commented on pkg/auth/static/auth.go:
```diff
@@ -22,11 +22,15 @@ func (s *StaticAuthenticator) Authenticate(w http.ResponseWriter, req *http.Requ
 
-func (s *StaticAuthenticator) LoginFunc(w http.ResponseWriter, req *http.Request)  {}
-func (s *StaticAuthenticator) LogoutFunc(w http.ResponseWriter, req *http.Request) {}
+func (s *StaticAuthenticator) LoginFunc(w http.ResponseWriter, req *http.Request) {
```
:+1:

stlaz commented on cmd/bridge/config/auth/authoptions.go:
```diff
@@ -252,12 +252,12 @@ func (c *completedOptions) getAuthenticator(
-		klog.Warning("running with AUTHENTICATION DISABLED!")
-
 		if c.BearerToken == "" {
+			klog.Warning("console is disabled -- no authentication method configured")
```
maybe underline that this is meant to be used for development only

TheRealJon replied:
I wanted to leave this vague since disabled auth can also represent an intermediate state when external OIDC is not fully configured. I didn't want to log "dev purpose only" lines in this case.

stlaz commented on cmd/bridge/config/auth/authoptions.go:
```diff
@@ -32,6 +33,7 @@ type AuthOptions struct {
 	ClientSecretFilePath string
 	CAFilePath           string
 	OCLoginCommand       string
+	BearerToken          string
```
I'd call it `StaticUserBearerToken`. Add a comment that explains that it'll override the authenticator and is supposed to be used for development purposes only.

stlaz commented on cmd/bridge/main.go:
```diff
@@ -561,15 +558,6 @@ func main() {
-		srv.Authenticator = static.NewStaticAuthenticator(auth.User{
-			Token: *fK8sAuthBearerToken,
-		})
-		srv.InternalProxiedK8SClientConfig.BearerToken = *fK8sAuthBearerToken
```
you'll want to keep this line somewhere

TheRealJon replied:
See https://github.com/openshift/console/pull/13713/files#diff-ae1fd264e062e64b8683f373f7ef6d59c5903f5471fa6dad594f9d505ac43681R223-R225

stlaz reviewed (commented):
> /label tide/merge-method-squash

I'd keep the original structure of the changes, there are too many and it's important that people in the future understand why each of the changes happened

stlaz commented:
/remove-label tide/merge-method-squash

TheRealJon commented:
@stlaz I've made some changes, mind taking another look?

TheRealJon commented:
@stlaz I've separated the auth type flag and PR feedback into separate commits. PTAL.

yanpzhan commented:
Checked on HCP cluster with the pr code, when AUTH type is OIDC and oidcClients is null, console operator won't degrade, console pod can run normally. Access console, it shows "Please configure authentication to use the web console." on the page.
@TheRealJon @stlaz The target branching is MAY 17, if there is not further update could the pr merge before that?

yanpzhan commented:
Checked on HCP cluster with the pr code, when AUTH type is OIDC and oidcClients is null, console operator won't degrade, console pod can run normally. Access console, it shows "Please configure authentication to use the web console." on the page.
@TheRealJon @stlaz The target branching is MAY 17, if there is not further update could the pr merge before that?

TheRealJon commented:
@yanpzhan There should be no further changes unless @stlaz has any further comments.

stlaz commented:
/cherry-pick release-4.16

stlaz commented:
/unhold

stlaz commented:
/jira refresh

stlaz commented:
/label acknowledge-critical-fixes

stlaz commented:
/label acknowledge-critical-fixes-only

stlaz commented:
/cherry-pick release-4.16

TheRealJon commented:
/retest
